docs/tutorial_basics.rst

```diff
@@ -231,10 +233,16 @@ ReFrame allows you to avoid this in several ways but the most compact is to defi
    :lines: 6-
 
 
-This is exactly the same test as the ``hello1.py`` except that it defines the ``lang`` parameter to denote the programming language to be used by the test.
-The :py:func:`~reframe.core.pipeline.RegressionTest.parameter` ReFrame built-in defines a new parameter for the test and will cause multiple instantiations of the test, each one setting the :attr:`lang` attribute to the actual parameter value.
-In this example, two tests will be created, one with ``lang='c'`` and another with ``lang='cpp'``.
-The parameter is available as an attribute of the test class and, in this example, we use it to set the extension of the source file.
+This test extends the ``hello1.py`` test by defining the ``lang`` parameter with the :py:func:`~reframe.core.pipeline.RegressionTest.parameter` built-in.
+This parameter will cause as many instantiations as parameter values available, each one setting the :attr:`lang` attribute to one single value.
+Hence, this example will create two test instances, one with ``lang='c'`` and another with ``lang='cpp'``.
+The parameter is available as an attribute of the test instance and, in this example, we use it to set the extension of the source file.
+However, at the class level, a test parameter holds all the possible values for itself, and this is only assigned a single value after the class is instantiated.
+In this case, the variable ``sourcepath`` must be set to ``f'hello.{self.lang}'``, but ``self.lang`` only acquires a specific value after the class is instantiated.
+For this reason the assignment of the ``sourcepath`` variable is deferred with a post-init hook, which will get executed after the class instantiation.
```

Comment:
I think it would make more sense to be `pre-compile` hook, wouldn't it?

Reply:
Yep, changed that. I've added some extra text here to explain a little bit more what hooks are.